# Traffic Sign Classification with Keras

## (See addendum at the bottom of this notebook - training the traffic sign neural net on the Cifar10 dataset)

Keras exists to make coding deep neural networks simpler. To demonstrate just how easy it is, you’re going to use Keras to build a convolutional neural network in a few dozen lines of code.

You’ll be connecting the concepts from the previous lessons to the methods that Keras provides.

## Dataset

The network you'll build with Keras is similar to the example in Keras’s GitHub repository that builds out a [convolutional neural network for MNIST](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py). 

However, instead of using the [MNIST](http://yann.lecun.com/exdb/mnist/) dataset, you're going to use the [German Traffic Sign Recognition Benchmark](http://benchmark.ini.rub.de/?section=gtsrb&subsection=news) dataset that you've used previously.

You can download pickle files with sanitized traffic sign data here:

In [2]:
from urllib.request import urlretrieve
from os.path import isfile
from tqdm import tqdm

class DLProgress(tqdm):
    last_block = 0

    def hook(self, block_num=1, block_size=1, total_size=None):
        self.total = total_size
        self.update((block_num - self.last_block) * block_size)
        self.last_block = block_num

if not isfile('train.p'):
    with DLProgress(unit='B', unit_scale=True, miniters=1, desc='Train Dataset') as pbar:
        urlretrieve(
            'https://s3.amazonaws.com/udacity-sdc/datasets/german_traffic_sign_benchmark/train.p',
            'train.p',
            pbar.hook)

if not isfile('test.p'):
    with DLProgress(unit='B', unit_scale=True, miniters=1, desc='Test Dataset') as pbar:
        urlretrieve(
            'https://s3.amazonaws.com/udacity-sdc/datasets/german_traffic_sign_benchmark/test.p',
            'test.p',
            pbar.hook)

print('Training and Test data downloaded.')

Training and Test data downloaded.


## Overview

Here are the steps you'll take to build the network:

1. Load the training data.
2. Preprocess the data.
3. Build a feedforward neural network to classify traffic signs.
4. Build a convolutional neural network to classify traffic signs.
5. Evaluate the final neural network on testing data.

Keep an eye on the network’s accuracy over time. Once the accuracy reaches the 98% range, you can be confident that you’ve built and trained an effective model.

In [3]:
import pickle
import numpy as np
import math

# Fix error with TF and Keras
import tensorflow as tf
tf.python.control_flow_ops = tf

print('Modules loaded.')

Modules loaded.


## Load the Data

Start by importing the data from the pickle file.

In [4]:
with open('train.p', 'rb') as f:
    data = pickle.load(f)

# TODO: Load the feature data to the variable X_train
X_train = data['features']
# TODO: Load the label data to the variable y_train
y_train = data['labels']

In [5]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
assert np.array_equal(X_train, data['features']), 'X_train not set to data[\'features\'].'
assert np.array_equal(y_train, data['labels']), 'y_train not set to data[\'labels\'].'
print('Tests passed.')

Tests passed.


## Preprocess the Data

1. Shuffle the data
2. Normalize the features using Min-Max scaling between -0.5 and 0.5
3. One-Hot Encode the labels

### Shuffle the data
Hint: You can use the [scikit-learn shuffle](http://scikit-learn.org/stable/modules/generated/sklearn.utils.shuffle.html) function to shuffle the data.

In [6]:
# TODO: Shuffle the data
from sklearn.utils import shuffle
X_train, y_train = shuffle(X_train, y_train)

In [7]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
assert X_train.shape == data['features'].shape, 'X_train has changed shape. The shape shouldn\'t change when shuffling.'
assert y_train.shape == data['labels'].shape, 'y_train has changed shape. The shape shouldn\'t change when shuffling.'
assert not np.array_equal(X_train, data['features']), 'X_train not shuffled.'
assert not np.array_equal(y_train, data['labels']), 'y_train not shuffled.'
print('Tests passed.')

Tests passed.


### Normalize the features
Hint: You solved this in [TensorFlow lab](https://github.com/udacity/CarND-TensorFlow-Lab/blob/master/lab.ipynb) Problem 1.

In [8]:
import cv2

# TODO: Normalize the data features to the variable X_normalized
X_normalized = X_train.astype(np.float32)/255.-0.5


In [9]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
assert math.isclose(np.min(X_normalized), -0.5, abs_tol=1e-5) and math.isclose(np.max(X_normalized), 0.5, abs_tol=1e-5), 'The range of the training data is: {} to {}.  It must be -0.5 to 0.5'.format(np.min(X_normalized), np.max(X_normalized))
print('Tests passed.')

Tests passed.


### One-Hot Encode the labels
Hint: You can use the [scikit-learn LabelBinarizer](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelBinarizer.html) function to one-hot encode the labels.

In [10]:
# TODO: One Hot encode the labels to the variable y_one_hot
from sklearn.preprocessing import LabelBinarizer
lb = LabelBinarizer()
y_one_hot = lb.fit_transform(y_train)

In [11]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
import collections

assert y_one_hot.shape == (39209, 43), 'y_one_hot is not the correct shape.  It\'s {}, it should be (39209, 43)'.format(y_one_hot.shape)
assert next((False for y in y_one_hot if collections.Counter(y) != {0: 42, 1: 1}), True), 'y_one_hot not one-hot encoded.'
print('Tests passed.')

Tests passed.


## Keras Sequential Model
```python
from keras.models import Sequential

# Create the Sequential model
model = Sequential()
```
The `keras.models.Sequential` class is a wrapper for the neural network model. Just like many of the class models in scikit-learn, it provides common functions like `fit()`, `evaluate()`, and `compile()`.  We'll cover these functions as we get to them.  Let's start looking at the layers of the model.

## Keras Layer
A Keras layer is just like a neural network layer.  It can be fully connected, max pool, activation, etc.  You can add a layer to the model using the model's `add()` function.  For example, a simple model would look like this:
```python
from keras.models import Sequential
from keras.layers.core import Dense, Activation, Flatten

# Create the Sequential model
model = Sequential()

# 1st Layer - Add a flatten layer
model.add(Flatten(input_shape=(32, 32, 3)))

# 2nd Layer - Add a fully connected layer
model.add(Dense(100))

# 3rd Layer - Add a ReLU activation layer
model.add(Activation('relu'))

# 4th Layer - Add a fully connected layer
model.add(Dense(60))

# 5th Layer - Add a ReLU activation layer
model.add(Activation('relu'))
```
Keras will automatically infer the shape of all layers after the first layer.  This means you only have to set the input dimensions for the first layer.

The first layer from above, `model.add(Flatten(input_shape=(32, 32, 3)))`, sets the input dimension to (32, 32, 3) and output dimension to (3072=32\*32\*3).  The second layer takes in the output of the first layer and sets the output dimenions to (100).  This chain of passing output to the next layer continues until the last layer, which is the output of the model.

## Build a Multi-Layer Feedforward Network

Build a multi-layer feedforward neural network to classify the traffic sign images.

1. Set the first layer to a `Flatten` layer with the `input_shape` set to (32, 32, 3)
2. Set the second layer to `Dense` layer width to 128 output. 
3. Use a ReLU activation function after the second layer.
4. Set the output layer width to 43, since there are 43 classes in the dataset.
5. Use a softmax activation function after the output layer.

To get started, review the Keras documentation about [models](https://keras.io/models/sequential/) and [layers](https://keras.io/layers/core/).

The Keras example of a [Multi-Layer Perceptron](https://github.com/fchollet/keras/blob/master/examples/mnist_mlp.py) network is similar to what you need to do here. Use that as a guide, but keep in mind that there are a number of differences.

In [12]:
from keras.models import Sequential
model = Sequential()
# TODO: Build a Multi-layer feedforward neural network with Keras here.

from keras.layers.core import Dense, Activation, Flatten

# 1st Layer - Add a flatten layer
model.add(Flatten(input_shape=(32, 32, 3)))

# 2nd Layer - Add a fully connected layer
model.add(Dense(128))

# 3rd Layer - Add a ReLU activation layer
model.add(Activation('relu'))

# 4th Layer - Add a fully connected layer
model.add(Dense(43))

# 5th Layer - Add a softmax activation layer
model.add(Activation('softmax'))


Using TensorFlow backend.


In [13]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten
from keras.activations import relu, softmax

def check_layers(layers, true_layers):
    assert len(true_layers) != 0, 'No layers found'
    for layer_i in range(len(layers)):
        assert isinstance(true_layers[layer_i], layers[layer_i]), 'Layer {} is not a {} layer'.format(layer_i+1, layers[layer_i].__name__)
    assert len(true_layers) == len(layers), '{} layers found, should be {} layers'.format(len(true_layers), len(layers))

check_layers([Flatten, Dense, Activation, Dense, Activation], model.layers)

assert model.layers[0].input_shape == (None, 32, 32, 3), 'First layer input shape is wrong, it should be (32, 32, 3)'
assert model.layers[1].output_shape == (None, 128), 'Second layer output is wrong, it should be (128)'
assert model.layers[2].activation == relu, 'Third layer not a relu activation layer'
assert model.layers[3].output_shape == (None, 43), 'Fourth layer output is wrong, it should be (43)'
assert model.layers[4].activation == softmax, 'Fifth layer not a softmax activation layer'
print('Tests passed.')

Tests passed.


## Training a Sequential Model
You built a multi-layer neural network in Keras, now let's look at training a neural network.
```python
from keras.models import Sequential
from keras.layers.core import Dense, Activation

model = Sequential()
...

# Configures the learning process and metrics
model.compile('sgd', 'mean_squared_error', ['accuracy'])

# Train the model
# History is a record of training loss and metrics
history = model.fit(x_train_data, Y_train_data, batch_size=128, nb_epoch=2, validation_split=0.2)

# Calculate test score
test_score = model.evaluate(x_test_data, Y_test_data)
```
The code above configures, trains, and tests the model.  The line `model.compile('sgd', 'mean_squared_error', ['accuracy'])` configures the model's optimizer to `'sgd'`(stochastic gradient descent), the loss to `'mean_squared_error'`, and the metric to `'accuracy'`.  

You can find more optimizers [here](https://keras.io/optimizers/), loss functions [here](https://keras.io/objectives/#available-objectives), and more metrics [here](https://keras.io/metrics/#available-metrics).

To train the model, use the `fit()` function as shown in `model.fit(x_train_data, Y_train_data, batch_size=128, nb_epoch=2, validation_split=0.2)`.  The `validation_split` parameter will split a percentage of the training dataset to be used to validate the model.  The model can be further tested with the test dataset using the `evaluation()` function as shown in the last line.

## Train the Network

1. Compile the network using adam optimizer and categorical_crossentropy loss function.
2. Train the network for ten epochs and validate with 20% of the training data.

In [15]:
# TODO: Compile and train the model here.
# Configures the learning process and metrics
model.compile('adam', 'categorical_crossentropy', ['accuracy'])

# Train the model
# History is a record of training loss and metrics
history = model.fit(X_normalized, y_one_hot, batch_size=128, nb_epoch=10, verbose=2, validation_split=0.2)


Train on 31367 samples, validate on 7842 samples
Epoch 1/10
4s - loss: 0.2409 - acc: 0.9331 - val_loss: 0.5919 - val_acc: 0.8471
Epoch 2/10
3s - loss: 0.2346 - acc: 0.9361 - val_loss: 0.3816 - val_acc: 0.8873
Epoch 3/10
3s - loss: 0.2133 - acc: 0.9419 - val_loss: 0.3748 - val_acc: 0.8966
Epoch 4/10
3s - loss: 0.1990 - acc: 0.9477 - val_loss: 0.3048 - val_acc: 0.9199
Epoch 5/10
3s - loss: 0.1962 - acc: 0.9467 - val_loss: 0.2636 - val_acc: 0.9299
Epoch 6/10
3s - loss: 0.1538 - acc: 0.9585 - val_loss: 0.3799 - val_acc: 0.8894
Epoch 7/10
3s - loss: 0.1684 - acc: 0.9521 - val_loss: 0.2630 - val_acc: 0.9291
Epoch 8/10
3s - loss: 0.1628 - acc: 0.9558 - val_loss: 0.2120 - val_acc: 0.9457
Epoch 9/10
3s - loss: 0.1490 - acc: 0.9591 - val_loss: 0.2300 - val_acc: 0.9371
Epoch 10/10
3s - loss: 0.1520 - acc: 0.9577 - val_loss: 0.2283 - val_acc: 0.9429


In [16]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.optimizers import Adam

assert model.loss == 'categorical_crossentropy', 'Not using categorical_crossentropy loss function'
assert isinstance(model.optimizer, Adam), 'Not using adam optimizer'
assert len(history.history['acc']) == 10, 'You\'re using {} epochs when you need to use 10 epochs.'.format(len(history.history['acc']))

assert history.history['acc'][-1] > 0.92, 'The training accuracy was: %.3f. It shoud be greater than 0.92' % history.history['acc'][-1]
assert history.history['val_acc'][-1] > 0.85, 'The validation accuracy is: %.3f. It shoud be greater than 0.85' % history.history['val_acc'][-1]
print('Tests passed.')

Tests passed.


## Convolutions
1. Re-construct the previous network
2. Add a [convolutional layer](https://keras.io/layers/convolutional/#convolution2d) with 32 filters, a 3x3 kernel, and valid padding before the flatten layer.
3. Add a ReLU activation after the convolutional layer.

Hint 1: The Keras example of a [convolutional neural network](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py) for MNIST would be a good example to review.

In [20]:
# TODO: Re-construct the network and add a convolutional layer before the flatten layer.
from keras.layers.convolutional import Convolution2D
model = Sequential()

# Add a convolution layer
model.add(Convolution2D(32, 3, 3, border_mode='valid', input_shape=(32, 32, 3)))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a flatten layer
model.add(Flatten())

# Add a fully connected layer
model.add(Dense(128))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a fully connected layer
model.add(Dense(43))

# Add a softmax activation layer
model.add(Activation('softmax'))


In [21]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten
from keras.layers.convolutional import Convolution2D

check_layers([Convolution2D, Activation, Flatten, Dense, Activation, Dense, Activation], model.layers)

assert model.layers[0].input_shape == (None, 32, 32, 3), 'First layer input shape is wrong, it should be (32, 32, 3)'
assert model.layers[0].nb_filter == 32, 'Wrong number of filters, it should be 32'
assert model.layers[0].nb_col == model.layers[0].nb_row == 3, 'Kernel size is wrong, it should be a 3x3'
assert model.layers[0].border_mode == 'valid', 'Wrong padding, it should be valid'

model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, nb_epoch=2, validation_split=0.2, verbose=2)
assert(history.history['val_acc'][-1] > 0.91), "The validation accuracy is: %.3f.  It should be greater than 0.91" % history.history['val_acc'][-1]
print('Tests passed.')

Train on 31367 samples, validate on 7842 samples
Epoch 1/2
31367/31367 [==============================] - 56s - loss: 1.1921 - acc: 0.6871 - val_loss: 0.4937 - val_acc: 0.8749

## Pooling
1. Re-construct the network
2. Add a 2x2 [max pooling layer](https://keras.io/layers/pooling/#maxpooling2d) immediately following your convolutional layer.

In [22]:
# TODO: Re-construct the network and add a pooling layer after the convolutional layer.

from keras.layers.pooling import MaxPooling2D

model = Sequential()

# Add a convolution layer
model.add(Convolution2D(32, 3, 3, border_mode='valid', input_shape=(32, 32, 3)))

# Add a max pooling layer
model.add(MaxPooling2D(pool_size=(2, 2), strides=None, border_mode='valid', dim_ordering='default'))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a flatten layer
model.add(Flatten())

# Add a fully connected layer
model.add(Dense(128))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a fully connected layer
model.add(Dense(43))

# Add a softmax activation layer
model.add(Activation('softmax'))

In [23]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten
from keras.layers.convolutional import Convolution2D
from keras.layers.pooling import MaxPooling2D

check_layers([Convolution2D, MaxPooling2D, Activation, Flatten, Dense, Activation, Dense, Activation], model.layers)
assert model.layers[1].pool_size == (2, 2), 'Second layer must be a max pool layer with pool size of 2x2'

model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, nb_epoch=2, validation_split=0.2, verbose=2)
assert(history.history['val_acc'][-1] > 0.91), "The validation accuracy is: %.3f.  It should be greater than 0.91" % history.history['val_acc'][-1]
print('Tests passed.')

Train on 31367 samples, validate on 7842 samples
Epoch 1/2
29s - loss: 1.5990 - acc: 0.5817 - val_loss: 0.7175 - val_acc: 0.7860
Epoch 2/2
30s - loss: 0.4922 - acc: 0.8755 - val_loss: 0.3601 - val_acc: 0.9171
Tests passed.


## Dropout
1. Re-construct the network
2. Add a [dropout](https://keras.io/layers/core/#dropout) layer after the pooling layer. Set the dropout rate to 50%.

In [24]:
# TODO: Re-construct the network and add dropout after the pooling layer.

from keras.layers.core import Dropout

model = Sequential()

# Add a convolution layer
model.add(Convolution2D(32, 3, 3, border_mode='valid', input_shape=(32, 32, 3)))

# Add a max pooling layer
model.add(MaxPooling2D(pool_size=(2, 2), strides=None, border_mode='valid', dim_ordering='default'))

# Add a dropout layer
model.add(Dropout(0.50))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a flatten layer
model.add(Flatten())

# Add a fully connected layer
model.add(Dense(128))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a fully connected layer
model.add(Dense(43))

# Add a softmax activation layer
model.add(Activation('softmax'))

In [25]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten, Dropout
from keras.layers.convolutional import Convolution2D
from keras.layers.pooling import MaxPooling2D

check_layers([Convolution2D, MaxPooling2D, Dropout, Activation, Flatten, Dense, Activation, Dense, Activation], model.layers)
assert model.layers[2].p == 0.5, 'Third layer should be a Dropout of 50%'

model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, nb_epoch=2, validation_split=0.2, verbose=2)
assert(history.history['val_acc'][-1] > 0.91), "The validation accuracy is: %.3f.  It should be greater than 0.91" % history.history['val_acc'][-1]
print('Tests passed.')

Train on 31367 samples, validate on 7842 samples
Epoch 1/2
32s - loss: 1.5555 - acc: 0.5916 - val_loss: 0.6299 - val_acc: 0.8556
Epoch 2/2
35s - loss: 0.5059 - acc: 0.8680 - val_loss: 0.3382 - val_acc: 0.9216
Tests passed.


## Optimization
Congratulations! You've built a neural network with convolutions, pooling, dropout, and fully-connected layers, all in just a few lines of code.

Have fun with the model and see how well you can do! Add more layers, or regularization, or different padding, or batches, or more training epochs.

What is the best validation accuracy you can achieve?

In [30]:
# TODO: Build a model

model = Sequential()

# Add a convolution layer
model.add(Convolution2D(6, 5, 5, border_mode='valid', input_shape=(32, 32, 3)))

# Add a max pooling layer
model.add(MaxPooling2D(pool_size=(2, 2), strides=None, border_mode='valid', dim_ordering='default'))

# Add a dropout layer
model.add(Dropout(0.50))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a convolution layer
model.add(Convolution2D(16, 5, 5, border_mode='valid'))

# Add a max pooling layer
model.add(MaxPooling2D(pool_size=(2, 2), strides=None, border_mode='valid', dim_ordering='default'))

# Add a dropout layer
model.add(Dropout(0.50))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a convolution layer
model.add(Convolution2D(400, 1, 1, border_mode='valid'))

# Add a dropout layer
model.add(Dropout(0.50))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a flatten layer
model.add(Flatten())

# Add a fully connected layer
model.add(Dense(128))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a fully connected layer
model.add(Dense(84))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a fully connected layer
model.add(Dense(43))

# Add a softmax activation layer
model.add(Activation('softmax'))

# TODO: Compile and train the model
model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, nb_epoch=12, validation_split=0.2, verbose=2)

Train on 31367 samples, validate on 7842 samples
Epoch 1/12
44s - loss: 2.5540 - acc: 0.3011 - val_loss: 1.0786 - val_acc: 0.6977
Epoch 2/12
42s - loss: 1.1662 - acc: 0.6302 - val_loss: 0.7015 - val_acc: 0.8141
Epoch 3/12
42s - loss: 0.9275 - acc: 0.7000 - val_loss: 0.5008 - val_acc: 0.8607
Epoch 4/12
42s - loss: 0.7887 - acc: 0.7426 - val_loss: 0.4376 - val_acc: 0.8596
Epoch 5/12
42s - loss: 0.7090 - acc: 0.7696 - val_loss: 0.3381 - val_acc: 0.9148
Epoch 6/12
42s - loss: 0.6539 - acc: 0.7843 - val_loss: 0.3145 - val_acc: 0.9086
Epoch 7/12
44s - loss: 0.6070 - acc: 0.7984 - val_loss: 0.2689 - val_acc: 0.9267
Epoch 8/12
18048s - loss: 0.5598 - acc: 0.8167 - val_loss: 0.2275 - val_acc: 0.9369
Epoch 9/12
11024s - loss: 0.5455 - acc: 0.8196 - val_loss: 0.2170 - val_acc: 0.9411
Epoch 10/12
47s - loss: 0.5069 - acc: 0.8334 - val_loss: 0.2297 - val_acc: 0.9389
Epoch 11/12
42s - loss: 0.4918 - acc: 0.8382 - val_loss: 0.1838 - val_acc: 0.9480
Epoch 12/12
41s - loss: 0.4872 - acc: 0.8388 - val_l

**Best Validation Accuracy:** 95.29%

## Testing
Once you've picked out your best model, it's time to test it.

Load up the test data and use the [`evaluate()` method](https://keras.io/models/model/#evaluate) to see how well it does.

Hint 1: The `evaluate()` method should return an array of numbers. Use the [`metrics_names`](https://keras.io/models/model/) property to get the labels.

In [33]:
# TODO: Load test data

with open('test.p', 'rb') as f:
    data = pickle.load(f)

X_test = data['features']
y_test = data['labels']

# TODO: Preprocess data & one-hot encode the labels
X_test_normalized = X_test.astype(np.float32)/255.-0.5

lb = LabelBinarizer()
y_test_one_hot = lb.fit_transform(y_test)

# TODO: Evaluate model on test data
test_score = model.evaluate(X_test_normalized, y_test_one_hot)


12630/12630 [==============================] - 9s      

In [37]:
print(model.metrics_names, test_score)

['loss', 'acc'] [0.33800812838857169, 0.90570071257963503]


**Test Accuracy:** 90.57%

## Summary
Keras is a great tool to use if you want to quickly build a neural network and evaluate performance.

## Addendum: Using the Traffic Sign Model for Cifar10 Dataset


In [1]:
from keras.datasets import cifar10
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

Using TensorFlow backend.


Untaring file...


In [3]:
import numpy as np
from sklearn.utils import shuffle
X_train, y_train = shuffle(X_train, y_train)

X_normalized = X_train.astype(np.float32)/255.-0.5

from sklearn.preprocessing import LabelBinarizer
lb = LabelBinarizer()
y_one_hot = lb.fit_transform(y_train)

In [4]:
from keras.layers.core import Dense, Activation, Flatten, Dropout
from keras.layers.convolutional import Convolution2D
from keras.models import Sequential
from keras.layers.pooling import MaxPooling2D
from keras.optimizers import Adam
import math
# Fix error with TF and Keras
import tensorflow as tf
tf.python.control_flow_ops = tf

In [5]:
model = Sequential()

# Add a convolution layer
model.add(Convolution2D(6, 5, 5, border_mode='valid', input_shape=(32, 32, 3)))

# Add a max pooling layer
model.add(MaxPooling2D(pool_size=(2, 2), strides=None, border_mode='valid', dim_ordering='default'))

# Add a dropout layer
model.add(Dropout(0.50))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a convolution layer
model.add(Convolution2D(16, 5, 5, border_mode='valid'))

# Add a max pooling layer
model.add(MaxPooling2D(pool_size=(2, 2), strides=None, border_mode='valid', dim_ordering='default'))

# Add a dropout layer
model.add(Dropout(0.50))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a convolution layer
model.add(Convolution2D(400, 1, 1, border_mode='valid'))

# Add a dropout layer
model.add(Dropout(0.50))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a flatten layer
model.add(Flatten())

# Add a fully connected layer
model.add(Dense(128))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a fully connected layer
model.add(Dense(84))

# Add a ReLU activation layer
model.add(Activation('relu'))

# Add a fully connected layer
model.add(Dense(10))

# Add a softmax activation layer
model.add(Activation('softmax'))

# TODO: Compile and train the model
model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, nb_epoch=12, validation_split=0.2, verbose=2)

Train on 40000 samples, validate on 10000 samples
Epoch 1/12
62s - loss: 1.8618 - acc: 0.3068 - val_loss: 1.8134 - val_acc: 0.3250
Epoch 2/12
60s - loss: 1.6473 - acc: 0.3865 - val_loss: 1.6766 - val_acc: 0.3969
Epoch 3/12
54s - loss: 1.5757 - acc: 0.4195 - val_loss: 1.5248 - val_acc: 0.4386
Epoch 4/12
61s - loss: 1.5187 - acc: 0.4441 - val_loss: 1.5052 - val_acc: 0.4662
Epoch 5/12
56s - loss: 1.4833 - acc: 0.4588 - val_loss: 1.4931 - val_acc: 0.4653
Epoch 6/12
53s - loss: 1.4550 - acc: 0.4713 - val_loss: 1.4432 - val_acc: 0.4723
Epoch 7/12
54s - loss: 1.4271 - acc: 0.4829 - val_loss: 1.4305 - val_acc: 0.4804
Epoch 8/12
60s - loss: 1.4112 - acc: 0.4885 - val_loss: 1.4192 - val_acc: 0.4814
Epoch 9/12
59s - loss: 1.3926 - acc: 0.4955 - val_loss: 1.4379 - val_acc: 0.4889
Epoch 10/12
53s - loss: 1.3791 - acc: 0.5006 - val_loss: 1.3457 - val_acc: 0.5171
Epoch 11/12
53s - loss: 1.3638 - acc: 0.5045 - val_loss: 1.3378 - val_acc: 0.5179
Epoch 12/12
53s - loss: 1.3568 - acc: 0.5137 - val_loss: 

In [6]:
X_test_normalized = X_test.astype(np.float32)/255.-0.5

lb = LabelBinarizer()
y_test_one_hot = lb.fit_transform(y_test)

test_score = model.evaluate(X_test_normalized, y_test_one_hot, verbose=2)
print(model.metrics_names, test_score)

['loss', 'acc'] [1.3839453199386598, 0.50190000000000001]


### Cifar10 dataset accuracy: 

# 50.19% 
(oh no!)